File: src/app/leetcode/p11/Solution.java
package app.leetcode.p11;

/**
 * MaxArea
 * for more, please refer to https://leetcode.com/problems/container-with-most-water/
 */
public class Solution {

    public int maxArea(int[] height) {
        int area = 0;
        int length = height.length;
        for (int i = 0; i < length; i++) {
            for (int j = i + 1; j < length; j++) {
                area = Math.max(area, (j - i) * Math.min(height[i], height[j]));
            }
        }
        return area;
    }

    public int maxArea2(int[] height) {
        int area = 0;
        int i = 0;
        int j = height.length - 1;
        while (i < j) {
            area = Math.max(area, (j - i) * Math.min(height[i], height[j]));
            if (height[i] < height[j]) {
                i++;
            } else {
                j--;
            }
        }

        return area;
    }

    public int maxArea3(int[] height) {
        int area = 0;
        int left = 0;
        int right = height.length - 1;
        while (left < right) {
            int minHeight = height[left] < height[right] ? height[left++] : height[right--];
            area = Math.max(area, (right - left + 1) * minHeight);
        }

        return area;
    }

    public static void main(String[] args) {
        int[] values = new int[] { 1, 8, 6, 2, 5, 4, 8, 3, 7 };
        int area = new Solution().maxArea2(values);

        System.out.println(area);
    }
}

File: src/app/leetcode/p11/Solution.swift

class Solution {
    func maxArea(_ height: [Int]) -> Int {
        var area = 0
        var i = 0
        var j = height.count - 1
        while i < j {
            area = max(area, (j - i) * min(height[i], height[j]))
            if height[i] < height[j] {
                i += 1
            } else {
                j -= 1
            }
        }
        return area
    }

    func maxArea1(_ height: [Int]) -> Int {
        var area = 0
        var left = 0
        var right = height.count - 1
        var h = 0
        while left < right {
            if height[left] < height[right] {
                h = height[left]
                left += 1
            } else {
                h = height[right]
                right -= 1
            }
            area = max(area, (right - left + 1) * h)
        }
        return area
    }
}
